crates/noirc_evaluator/src/ssa_refactor/ir/post_order.rs

```diff
@@ -0,0 +1,166 @@
+//! The post-order for a given function represented as a vector of basic block ids.
+//!
+//! This ordering is beneficial to the efficiency of various algorithms, such as those for dead
+//! code elimination and calculating dominance trees.
+
+use std::collections::HashSet;
+
+use crate::ssa_refactor::ir::{basic_block::BasicBlockId, function::Function};
+
+/// DFT stack state marker for computing the cfg post-order.
+enum Visit {
+    First,
+    Last,
+}
+
+pub(crate) struct PostOrder(Vec<BasicBlockId>);
+
+impl PostOrder {
+    pub(crate) fn as_slice(&self) -> &[BasicBlockId] {
+        self.0.as_slice()
+    }
+}
+
+impl PostOrder {
+    /// Allocate and compute a function's block post-order. Pos
+    pub(crate) fn with_function(func: &Function) -> Self {
+        PostOrder(Self::compute_post_order(func))
+    }
+
+    // Computes the post-order of the function by doing a depth-first traversal of the
+    // function's entry block's previously unvisited children. Each block is sequenced according
+    // to when the traversal exits it.
+    fn compute_post_order(func: &Function) -> Vec<BasicBlockId> {
+        let mut stack = vec![(Visit::First, func.entry_block())];
+        let mut visited: HashSet<BasicBlockId> = HashSet::new();
+        let mut post_order: Vec<BasicBlockId> = Vec::new();
+
+        while let Some((visit, block_id)) = stack.pop() {
+            match visit {
+                Visit::First => {
+                    if !visited.contains(&block_id) {
+                        // This is the first time we pop the block, so we need to scan its
+                        // successors and then revisit it.
+                        visited.insert(block_id);
+                        stack.push((Visit::Last, block_id));
+                        // Stack successors for visiting. Because items are taken from the top of the
+                        // stack, we push the item that's due for a visit first to the top.
+                        for successor_id in
+                            func.dfg[block_id].successors().collect::<Vec<_>>().into_iter().rev()
```

Comment:
`successors` should already be an iterator so we should be able to do:
```rs
func.dfg[block_id].successors().rev()
```

Reply:
I had to add `DoubleEndedIterator` to `successors` return type.